### PR TITLE
Add IIS Tool contrib package

### PR DESCRIPTION
**Pull request by bknueven**
## Summary/Motivation:
This formalizes a workflow I've used for a while to compute the IIS of an infeasible Pyomo model using the persistent solver interface of a commercial solver.

## Changes proposed in this PR:
- Add Infeasible Irreducible System (IIS) Tool to contrib.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/contrib/iis/iis.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    available_solvers = [
+        s
+        for s, sp in zip(_supported_solvers, _supported_solvers_persistent)
+        if SolverFactory(sp).available(exception_flag=False)
+    ]
+
+    if solver is None:
+        if len(available_solvers) == 0:
+            raise RuntimeError("Could not find a solver to use")
+        solver = _available_solvers[0]
```
Comment (michaelbynum) [inline pyomo/contrib/iis/iis.py:41]: is _available_solvers defined? It looks like there is no underscore above.

File: pyomo/contrib/iis/iis.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        print(f"Using solver {solver}")
```
Comment (michaelbynum) [inline pyomo/contrib/iis/iis.py:42]: I would replace all print statements with logging statements so they can be turned off if needed.

File: pyomo/contrib/iis/tests/test_iis.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    m.c2 = pyo.Constraint(expr=m.y <= -100.0 * m.x)
+    m.c3 = pyo.Constraint(expr=m.x >= 0.5)
+
+    m.o = pyo.Objective(expr=-m.y)
+
+    return m
+
+
+class TestIIS(unittest.TestCase):
+    @unittest.skipUnless(
```
Comment (michaelbynum) [inline pyomo/contrib/iis/tests/test_iis.py:24]: The tests look great. I would just recommend adding one more test for the case where the solver is not specified:

```
  @unittest.skipUnless(
         (pyo.SolverFactory("cplex_persistent").available(exception_flag=False) or
          pyo.SolverFactory("gurobi_persistent").available(exception_flag=False) or
          pyo.SolverFactory("xpress_persistent").available(exception_flag=False)),
         "Persistent solver not available",
     )
     def test_write_iis_any_solver(self):
         _test_iis(None)
```

I think this should work without changing any of the other test code???

**Review by michaelbynum — CHANGES_REQUESTED** [review]
This is awesome! Thanks, @bknueven!

File: pyomo/contrib/iis/iis.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+    iis = IISFactory(solver)
+    iis.compute()
+    iis_file_name = iis.write(iis_file_name)
+    print(f"IIS written to {iis_file_name}")
+    return iis_file_name
+
+
+def _remove_suffix(string, suffix):
+    if suffix in string:
```
Comment (jsiirola) [inline pyomo/contrib/iis/iis.py:65]: This should be `if string.ends_with(suffix):`

File: pyomo/contrib/iis/iis.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+            return file_name
+        else:
+            return _remove_suffix(file_name, ".lp") + ".lp"
+
+
+_solver_map = {
+    "cplex_persistent": CplexConflict,
+    "gurobi_persistent": GurobiIIS,
+    "xpress_persistent": XpressIIS,
+}
```
Comment (jsiirola) [inline pyomo/contrib/iis/iis.py:126]: @michaelbynum, How challenging would it be to add support for the APPSI interfaces as well?

**Review by jsiirola — COMMENTED** [review]
Thanks, @bknueven! Some questions / requests, but overall, this is great!

File: pyomo/contrib/iis/iis.py (hunk 5 above)
Comment (michaelbynum) [reply pyomo/contrib/iis/iis.py:126]: Not hard for Gurobi. I still have not refactored Appsi's interface to cplex to use the python bindings.

**Comment by bknueven** [comment]
Thanks for the thorough review @michaelbynum and @jsiirola. I believe I have addressed all of your comments.

I noticed incidentally that this does not quite work for Cplex 12.10 (and probably not the whole 12.x series). I may have a go at correcting that in the next few days.

I've tested locally the oldest release of Xpress available on pypi, and as for Gurobi I think its IIS API has been the same since at least version 5.6.

**Review by michaelbynum — APPROVED** [review]
I think it is okay if we do not support this functionality for older versions of solvers.

**Review by jsiirola — APPROVED** [review]

**Pull request merged**